File: Data Structures & Algorithms/replace-elements-with-greatest-element-on-right-side/submission-0.cpp
class Solution {
public:
    vector<int> replaceElements(vector<int>& arr) {
        int n = arr.size(); 
        int max = -1; 
        vector<int> res(n); 

        // iterate backwards
        for(int i = n-1; i >= 0; i--) {
            // push current max
            res[i] = max; 

            // check if current value for max
            max = std::max(arr[i], max); 
            
        }

        return res;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cs
public class Solution {
    public int CharacterReplacement(string s, int k) {
        // Check edge cases
        if(s.Length == 0) return 0; 
        
        // Set up a dictionary to count frequencies
        Dictionary<char, int> freq = new Dictionary<char, int>(); 
        int res = 0, l = 0, maxf = 0; 
        // Populate the dictionary
        for(int r = 0; r < s.Length; r++) {
            if(!freq.ContainsKey(s[r])) freq[s[r]] = 0; 

            freq[s[r]]++;
            // get the highest frequency character
            maxf = Math.Max(maxf, freq[s[r]]); 
            
            // iterate while  window size minus max frequency is greater than k 
            while((r - l + 1) - maxf > k) {
                // adjust the count for shrinking window 
                freq[s[l]]--;
                l++;
            }
            res = Math.Max(res, r - l +1); 
        }

        return res; 
    }
}
